src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -165,18 +164,31 @@ public override void AdvanceTo(SequencePosition consumed)
 
         public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
         {
-            if (_inputLength == 0)
+            var examinedLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, examined).Length;
+            var consumedLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+
+            if (_readCompleted)
             {
+                _readResult = new ReadResult(_readResult.Buffer.Slice(consumed, _readResult.Buffer.End), isCanceled: false, _readCompleted);
+                if (!_firstAdvance)
+                {
+                    _context.Input.AdvanceTo(consumed, examined);
+                    _firstAdvance = true;
+                }
                 return;
             }
 
-            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
-
-            _inputLength -= dataLength;
-
             _context.Input.AdvanceTo(consumed, examined);
 
-            OnDataRead(dataLength);
+            var newlyExamined = examinedLength - _totalExaminedInPreviousReadResult;
+            if (newlyExamined > 0)
```

Comment:
Did you confirm if examined can go backwards? Or does Advance throw?

Reply:
Advance doesn't throw.

Reply:
Examined can't go backwards.